### ISSUE TITLE
Compiling to categories for Agda

### ISSUE THREAD

**Issue by conal** [conal/calculating-compilers-agda#3]
I’m loving my experience with Agda so far. Can we reproduce and improve on [compiling to categories](http://conal.net/papers/compiling-to-categories/) in Agda? The idea would be to program directly in Agda, convert to categorical form, and re-interpret in different categories. This stack machine project could be one of many applications.

One challenge is how translate from Agda’s more powerfully typed lambda calculus to categorical terms.

**Comment by conal** [comment]
@ahmadsalim Thoughts?

**Comment by ahmadsalim** [comment]
@conal This would require I think working with Agda's reflection mechanism. I am not very familiar with it (as in having written code), but I think it should definitely be possible!

**Comment by JacquesCarette** [comment]
[Hopefully you don't mind the intrusion - I've been following this work, and find it quite interesting.]

Rather than "programming directly in Agda", which would then force you to use reflection [which has all sorts of limitations that will drive you crazy], I would probably choose to program in a language that embeds in Agda. Creating well-typed DSLs in Agda is easy.

I would also suggest that you use agda-categories rather than rolling your own. All the machinery needed for 'compiling to categories' already exists.

**Comment by conal** [comment]
@ahmadsalim and @jacquescarette Thanks for your suggestions, including reflection. I pursued a similar (I assume) path with Template Haskell several years ago. Perhaps Agda reflection works better, but struggling with its limitations sounds unappealing.

DSL embedding was unsatisfactory as well, since it didn’t allow the “overloading” of lambda, variables, and application that compiling-to-categories (CtoC) enables, along with its often-ideal programming interface (none beyond the host language) and inspection/analysis ability. (I did mostly DSL embeddings before the CtoC alternative.)

Implementing CtoC for Haskell required a compiler plugin. Is there any such mechanism for Agda?

**Comment by ahmadsalim** [comment]
@conal As far as I know, there is no compiler plugin mechanism for Agda.

I think what @JacquesCarette suggests is a great idea! E.g., https://plfa.github.io/Lambda/ has some way to fake lambda syntax, which is not as nice as the built-in one but Agda is sufficiently strongly typed to allow statically checking e.g., variable scoping.

**Comment by conal** [comment]
BTW, my intent is to write regular functional programs (in Haskell or Agda) and then give them principled but nonstandard interpretations such as differentiable functions (e.g., for machine learning), incremental functions (for efficient response to incrementally changing inputs), and massively parallel circuits. I don’t imagine it would be enjoyable to write useful, nontrivial programs in fake lambda syntax; do you?

**Comment by ahmadsalim** [comment]
No, you are right. If you want to write fully fledged programs, then it is not nice to use this fake lambda syntax 🙂

**Comment by ahmadsalim** [comment]
Maybe https://github.com/alhassy/gentle-intro-to-reflection is useful to get a feel for how good Agda reflection is for the purpose?

**Comment by JacquesCarette** [comment]
First, let me recapitulate my understanding of the project, to make sure we're on the same page.  [There will be a second post with 'new' content]. We know (from Scott in 1980 - https://ncatlab.org/nlab/show/lambda-calculus) that the simply typed lambda calculus (STLC) is the internal language of CCCs. This means that we can use the syntax of STLC to denote anything that can be interpreted in a CCC.  So the syntax of STLC is natively very polymorphic.  Of course other calculi [are known](https://ncatlab.org/nlab/show/relation+between+type+theory+and+category+theory) to correspond to the internal language of various categories.

Then one can indeed interpret such terms in a variety of places, as soon as a CCC structure is exhibited.

Section 2 of [compiling to categories](http://conal.net/papers/compiling-to-categories/compiling-to-categories.pdf) indeed shows that **Hask** is a rather good approximation to just that. [Approximation because of lazyness and other oddities, but it's close enough to not overly worry about].

Section 3 then gives a syntax-directed translation from STLC to a point-free form. This is somewhat difficult because the axiomatization of Closed doesn't furnish all of the extra goodies that normally are there -- see Categories.Category.CartesianClosed and Categories.Object.Exponential in the agda-categories project to see some of them.

Section 4 says that CC Functors, i.e. Functors that preserve the CC structure) are darn useful to transport things around.

Section 5 says that's it's a lot of work to engineer GHC to do this, since `->` is so internal to it. But it's possible.  [This would be easier in Scala as they do allow you to overload everything].

Section 6 says that we really want smaller Hom sets than just those of **Hask**. Luckily, some sub-categories are indeed available.

Section 7 shows that a certain Graph category looks to be CCC [skipping the proofs...], and that Syn (which is really close to a codiscrete category, since all Homs can be inhabited) is a CCC as well. AD sneaks in because the sub-category with just linear maps is a CCC as well [though I've never seen a full proof of that], etc.

Section 8 brings to our attention that other structure than CCC can be useful in our internal language. Interestingly, cocartesian doesn't show up on the nLab page, nor does distributive.  Though perhaps these are also "Rig Categories", about which I've written about.

Section 9 deals wtih representation issues, and 10 with implementation issues [which are non-trivial].

Another place where the usefulness of "type theory" as internal syntax for a variety of categories is really well explained is Mike Shulman's [superb notes](http://mikeshulman.github.io/catlog/catlog.pdf) on categorical logic from a categorical point of view.  Despite the title, it's as much about type theory as about category theory!

=============

So the idea would then be to use a system in which it is possible to write in a sufficiently nice syntax (a la Haskell), but to get one's hands on the resulting terms, to be able to re-target them to other settings.

I think that can be done in Agda, without the need for reflection -- assuming HOAS is ok.  [But I'm not 100% sure].  Otherwise it gets tricky.

**Comment by JacquesCarette** [comment]
Note that if that guide to Agda reflection is 'not enough', ask -- Musa, the author, is my Ph.D. student!  He wrote this while he was investigating if reflection was enough for his work [it's not - it can't deal with generating new names].

**Comment by JacquesCarette** [comment]
Having said that, if you **don't** need fresh name generation, then Agda reflection is very powerful indeed.  You can get at the full syntax of a term. Great for analysis, or proof by reflection. Not so great when you're trying to do 'generation' (say of the Deriving kind).

**Comment by ahmadsalim** [comment]
@JacquesCarette Thanks for the clarification 😄. I like Musa's tutorial!

HOAS does not work well with Agda (with my experience), and you need things like hereditary substitution, which is a pain to implement and use.

If the issue is (only) that Agda's fresh name generation is lacking, we could always see if it is easy to improve the Agda source code. They are usually very welcoming to feedback and it seems like a simple feature to add (compared to a compiler plug-in architecture).

**Comment by JacquesCarette** [comment]
The [following issue](https://github.com/agda/agda/issues/3699) will definitely be relevant.

**Comment by conal** [comment]
> if you **don’t** need fresh name generation, then Agda reflection is very powerful indeed.

I did rely on fresh name generation for the GHC plugin.

**Comment by JacquesCarette** [comment]
You might need to dig in to Agda itself then.  But my work is likely to need that too, so a little coordination might save us both time.

You could proceed as @alhassy has done for [his first prototype](https://github.com/alhassy/next-700-module-systems): write it in elisp!
